### DISCUSSION TITLE
A bit lost with Black Box MOGA

### DISCUSSION THREAD

**Discussion by colvinem** [snl-dakota/dakota#159] [Q&A]
I'm fighting with a multi-objective optimization problem where I use a system interface with a python script (which the servers have decided will only work with python2 and it is just easier to work around that than to try to get it to use python3 at this point) that takes the input from the parameters file (I'm skipping an external preprocessor and just doing it myself), generates a Serpent input, runs that, spits the flux out into an ORIGEN input, then pulls a couple of the results out to put into the results file. I've finally gotten rid of everything that needs python3 to run and both Serpent and ORIGEN are running just fine. But I'm still running into issues where I get this error: `Error: cannot open results file "results.out.1" for evaluation 1`

I have no clue why. The results.out.1 file is clearly there and looks right (the name is not hard-coded - it pulls it from the input argument that DAKOTA passes to the Python script). This is what I'm doing to write it:
```
resp_file = open('temp.out','w')
resp_file.write(str(var1)+' var1\n')
resp_file.write(str(var2)+' var2\n')
resp_file.close()

runcopy = 'mv temp.out '+str(resfile)+' &'
os.system(runcopy)
```
The file permissions are `rw-rw----` which I think should be fine? This is what I've got in the DAKOTA input:
```
interface
    system
        analysis_driver = "python2 scriptb0.py"
        parameters_file = "params.in"
        results_file = "results.out"
        file_tag file_save

responses
    num_objective_functions = 2
    descriptors = "var1" "var2"
    no_gradients
    no_hessians
```
I feel like I'm probably missing something obvious, but I'm at a loss.

**Comment by thenectarine** [comment]
Can you share the full Dakota input file and Python script? If not, can you provide a simple test case that reproduces the error?

The error means that Dakota does not find the file in the expected location. This could either be because the black box function or Python script failed during execution, so the results file is absent, OR the results file was written to a different directory that Dakota was not expecting. In very rare occasions, I did encounter the same error when pre existing result files were present during the Dakota run, but this was only reproducible once in a blue moon. In this situation, I simply deleted the old result files to avoid the issue. While typing this, I was wondering if a separate process on Linux was temporarily accessing and restricting old result files, which prevented Dakota from properly managing the files. You mentioned the results file does exist and does have results/values in the file, so perhaps there an issue with the working directory?

What's the working directory of the Dakota run? Is it the same working directory as the Python execution?

If the working directories are not identical, this might explain the issue. If Dakota expects results.out.1 in `/home/user_a/dir_a`, but Python writes out the file to `/home/user_a/dir_b`, this is a problem. 

Normally the working directories are identical, but if your input files and scripts are sufficiently complex, there might be a statement that changes the working directory. I believe Dakota does have some keywords to modify the working directory, e.g. work_directory. Python allows you to change the working directory with os.chdir(). 

I rarely do work one level down from the root level, e.g. `/temp`, `/scratch`, because some OSs limit the permissions in these directories. I typically work 2 or 3 levels down from the root level, e.g. `/home/user_a/Downloads`. Generally, anything within the user's directory should be OK, but anything above could be problematic.

By the way, are you running Dakota, Python and the black box functions all on the same operating system? If you are running everything on a local OS, then disregard this paragraph. If you are running Dakota on a local operating system, but the black box function is being executed on a remote operating system, care must be taken to ensure the results file is available on the local OS for Dakota's use.

I tested multi objective optimization on my end with the files below. This simple test case worked for me. This simple test case should also work for you, but make sure to change the path `/opt/dakota/test/cyl_head` to match your location of Dakota's test files.

**Working Example**

dakota_multiobj3.in

```
method,
       moga     						#s3
	  seed = 10983                              		#s2,#s3
	  max_function_evaluations = 10             		#s2,#s3
	  initialization_type unique_random            		#s2,#s3
	  mutation_type offset_normal				#s2,#s3
	    mutation_scale = 0.1				#s2,#s3
	  final_solutions = 5	     				#s3
	  crossover_type            				#s3
	    multi_point_parameterized_binary = 2        	#s3
	    crossover_rate = 0.8            			#s3
	  fitness_type domination_count            		#s3
	  replacement_type below_limit = 6            		#s3
	    shrinkage_percentage = 0.9            		#s3
	  convergence_type metric_tracker            		#s3
	    percent_change = 0.05 num_generations = 10          #s3

variables,
	continuous_design = 2
	  initial_point    1.8    1.0
	  upper_bounds     2.164  4.0
	  lower_bounds     1.5    0.0
	  descriptors'intake_dia' 'flatness'

interface,
	  #analysis_driver=  '/opt/dakota/test/cyl_head'
      analysis_drivers  'python2 scriptb0.py'
        system
        parameters_file = "params.in"
        results_file = "results.out"
        file_tag  file_save

responses,
	objective_functions = 4
	  multi_objective_weights = .7 .1 .1 .1			#s0,#s1,#s2
	  sense = "min" "max" "max" "max"	       	   	#s1,#s2,#s3
	no_gradients
	no_hessians
```

scriptb0.py

```
import sys
import subprocess
import os

# Look at working directory
print(os.getcwd())

# Get arguments
print(sys.argv)
resfile = sys.argv[2]

# Run black box function and write out results
subprocess.call(['/opt/dakota/test/cyl_head', sys.argv[1], 'temp.out'])

# Rename the results file
runcopy = 'mv temp.out '+str(resfile)+' &'
os.system(runcopy)
```
